trackintel/preprocessing/positionfixes.py

```diff
@@ -173,187 +177,335 @@ def generate_triplegs(positionfixes, staypoints=None, *args, **kwargs):
+            positionfixes_user_this = ret_pfs.loc[ret_pfs['user_id'] == user_id_this].sort_values(
+                'tracked_at')  # this is no copy
 
-                # get the last posfix of the first staypoint
-                index_first_posfix_tl = pfs[pfs.staypoint_id == stp1['id']].index[-1]
-                position_first_posfix_tl = pfs.index.get_loc(index_first_posfix_tl)
+            # Case 1: Staypoints exist and are connected to positionfixes by user id
+            if case == 1:
+                generated_triplegs.extend(_triplegs_between_staypoints_case1(positionfixes_user_this, staypoints,
+                                                                             user_id_this))
 
-                # get first posfix of the second staypoint
-                index_last_posfix_tl = pfs[pfs.staypoint_id == stp2['id']].index[0]
-                position_last_posfix_tl = pfs.index.get_loc(index_last_posfix_tl)
+            # Case 2: Staypoints exist but there is no user_id given
+            elif case == 2:
+                generated_triplegs.extend(_triplegs_between_staypoints_case2(positionfixes_user_this, staypoints,
+                                                                             user_id_this))
 
-                pfs_tripleg = pfs.iloc[position_first_posfix_tl:position_last_posfix_tl + 1]
+            # case 3: Only positionfixes with staypoint id for tripleg generation
+            elif case == 3:
+                generated_triplegs.extend(_triplegs_between_staypoints_case3(positionfixes_user_this, user_id_this))
 
-                # include every positionfix that brings you closer to the center 
-                # of the staypoint
+        # create tripleg dataframe
+        if len(generated_triplegs) == 0:
+            ret_tpls = gpd.GeoDataFrame(columns=['id', 'user_id', 'started_at', 'finished_at', 'geom', 'pfs_ids'],
+                                        geometry='geom', crs=ret_pfs.crs)
+        else:
+            ret_tpls = gpd.GeoDataFrame(generated_triplegs, geometry='geom', crs=ret_pfs.crs)
 
-                started_at = pfs_tripleg['tracked_at'].iloc[0]
-                finished_at = pfs_tripleg['tracked_at'].iloc[-1]
+        ret_tpls.index.name = 'id'
 
-                coords = list(pfs_tripleg.geometry.apply(lambda r: (r.x, r.y)))
+        # assign tripleg_id to positionfixes
+        ret_pfs['tripleg_id'] = np.nan
 
-                if len(coords) > 1:
-                    generated_triplegs.append({
-                        'id': curr_tripleg_id,
-                        'user_id': user_id_this,
-                        'started_at': started_at,  # pfs_tripleg['tracked_at'].iloc[0],
-                        'finished_at': finished_at,  # pfs_tripleg['tracked_at'].iloc[-1],
-                        'geom': LineString(coords)
-                    })
-                    curr_tripleg_id += 1
+        for tripleg_id, tripleg in ret_tpls.iterrows():
```

Comment:
Here I would assume iterations are quite slow, is it possible to use join or merge to directly assign the ids? (pfs on index and tpls on pfs_id). 
Check Line 97 in `trackintel\preprocessing\positionfixes.py` or Line 135-138 in `trackintel\preprocessing\triplegs.py`

Reply:
👍